c++/src/io/Cache.hh

```diff
@@ -0,0 +1,208 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <orc/MemoryPool.hh>
+#include <orc/OrcFile.hh>
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <future>
+#include <utility>
+#include <vector>
+
+namespace orc {
+  class InputStream;
+
+  struct CacheOptions {
+    /// The maximum distance in bytes between two consecutive
+    /// ranges; beyond this value, ranges are not combined
+    uint64_t holeSizeLimit = 8192;
+
+    /// The maximum size in bytes of a combined range; if
+    /// combining two consecutive ranges would produce a range of a
+    /// size greater than this, they are not combined
+    uint64_t rangeSizeLimit = 32 * 1024 * 1024;
+  };
+
+  struct ReadRange {
+    uint64_t offset;
+    uint64_t length;
+
+    ReadRange() = default;
+    ReadRange(uint64_t offset, uint64_t length) : offset(offset), length(length) {}
+
+    friend bool operator==(const ReadRange& left, const ReadRange& right) {
+      return (left.offset == right.offset && left.length == right.length);
+    }
+    friend bool operator!=(const ReadRange& left, const ReadRange& right) {
+      return !(left == right);
+    }
+
+    bool contains(const ReadRange& other) const {
+      return (offset <= other.offset && offset + length >= other.offset + other.length);
+    }
+  };
+
+  struct ReadRangeCombiner {
+    std::vector<ReadRange> coalesce(std::vector<ReadRange> ranges) const {
+      if (ranges.empty()) {
+        return ranges;
+      }
+
+      // Remove zero-sized ranges
+      auto end = std::remove_if(ranges.begin(), ranges.end(),
+                                [](const ReadRange& range) { return range.length == 0; });
+      // Sort in position order
+      std::sort(ranges.begin(), end, [](const ReadRange& a, const ReadRange& b) {
+        return a.offset != b.offset ? a.offset < b.offset : a.length > b.length;
+      });
+
+      // Remove ranges that overlap 100%
+      end = std::unique(ranges.begin(), end, [](const ReadRange& left, const ReadRange& right) {
+        return left.contains(right);
+      });
+      ranges.resize(end - ranges.begin());
+
+      // Skip further processing if ranges is empty after removing zero-sized ranges.
+      if (ranges.empty()) {
+        return ranges;
+      }
+
+#ifndef NDEBUG
+      for (size_t i = 0; i < ranges.size() - 1; ++i) {
+        const auto& left = ranges[i];
+        const auto& right = ranges[i + 1];
+        assert(left.offset < right.offset);
+        assert(!left.contains(right));
+      }
+#endif
+
+      std::vector<ReadRange> coalesced;
+
+      auto itr = ranges.begin();
+      // Ensure ranges is not empty.
+      assert(itr <= ranges.end());
```

Comment:
Remove this

Reply:
But assert check the conditions in debug mode, keeping it doesn't influence release version